README.md
# MarketNeutralTools

history.txt

# 2021-12-27 20:20:49.377916
+MSFT

# 2021-12-27 20:20:51.131358
+AAPL

# 2021-12-27 20:20:52.793346
+SPY

# 2021-12-27 20:21:57.360849
+Usman

# 2021-12-27 20:21:58.732008
+Shaizah

# 2021-12-27 20:22:02.503624
+Usman

# 2021-12-27 20:26:14.683057
+AAPL

# 2021-12-27 20:26:16.453441
+MSFT

# 2021-12-27 20:26:19.959170
+SPY

# 2021-12-27 20:26:34.866543
+Usman

# 2021-12-27 20:26:36.322276
+Shaizah

# 2021-12-27 20:26:37.255894
+SPY

# 2021-12-27 20:26:39.154574
+AAPL

# 2021-12-27 20:26:40.032297
+MSFT

# 2021-12-27 20:30:57.890803
+2

# 2021-12-27 20:30:59.667573
+1

# 2021-12-27 20:31:01.763645
+Hello

# 2021-12-27 20:32:28.944257
+MSFT

# 2021-12-27 20:32:30.719136
+AAPL

# 2021-12-27 20:34:06.726208
+MSFT

# 2021-12-27 20:36:30.691022
+AAPL

# 2021-12-28 01:12:35.989597
+MSFT

# 2021-12-28 13:04:11.818757
+SPY

# 2021-12-28 14:09:49.370644
+AAPL

# 2021-12-28 14:40:07.997718
+SPY

# 2021-12-28 14:40:18.852161
+AAPL

# 2021-12-28 14:41:32.364003
+MSFT

# 2021-12-28 14:41:46.231381
+SPY

# 2021-12-28 14:52:55.122244
+AAPL

# 2021-12-28 14:53:32.166970
+MSFT

# 2021-12-28 14:53:55.785259
+SPY

# 2021-12-28 14:57:41.563979
+AAPL

# 2021-12-28 14:59:24.561784
+SPY

# 2021-12-28 14:59:44.837194
+AAPL

# 2021-12-28 15:02:11.896278
+MSFT

# 2021-12-28 15:26:47.112263
+AAPL

# 2021-12-28 15:26:52.716174
+MSFT

# 2021-12-28 15:31:32.978069
+AAPL

# 2021-12-28 15:40:27.313182
+MSFT

# 2021-12-28 15:41:51.392670
+AAPL

# 2021-12-28 15:42:02.212259
+MSFT

# 2021-12-28 15:42:30.987543
+AAPL

# 2021-12-28 15:45:22.610126
+MSFT

# 2021-12-28 15:45:27.276242
+AAPL

# 2021-12-28 15:46:07.191699
+SPY

# 2021-12-28 15:46:53.087056
+MSFT

# 2021-12-28 15:46:56.486410
+AAPL

# 2021-12-28 15:47:55.735897
+MSFT

# 2021-12-28 15:47:59.279385
+AAPL

# 2021-12-28 15:53:29.281082
+MSFT

# 2021-12-28 15:53:33.449676
+AAPL

# 2021-12-28 15:54:01.897077
+MSFT

# 2021-12-28 15:55:21.675834
+AAPL

# 2021-12-28 15:58:32.736243
+MSFT

# 2021-12-28 15:59:49.237220
+AAPL

# 2021-12-28 15:59:52.226931
+MSFT

# 2021-12-28 16:00:17.615686
+AAPL

# 2021-12-28 16:02:24.964885
+MSFT

# 2021-12-28 16:02:28.301224
+AAPL

# 2021-12-28 16:03:14.724784
+msft

# 2021-12-28 16:03:18.730183
+aap

# 2021-12-28 16:03:46.800128
+msft

# 2021-12-28 16:05:24.925772
+MSFT

# 2021-12-28 16:05:30.546783
+AAPL

# 2021-12-28 16:06:27.748477
+MSFT

# 2021-12-28 16:06:40.677626
+SPY

# 2021-12-28 16:09:36.272860
+AAPL

# 2021-12-28 16:11:49.293384
+MSFT

# 2021-12-28 16:15:54.303172
+AAPL

# 2021-12-28 16:17:19.298365
+MSFT

# 2021-12-28 16:31:15.704754
+AAPL

# 2021-12-28 19:59:42.760933
+MSFT

# 2021-12-28 20:00:23.453179
+AAPL

# 2021-12-28 20:02:10.652172
+MSFT

# 2021-12-28 20:04:17.692475
+AAPL

# 2021-12-28 20:04:57.439942
+MSFT

# 2021-12-30 21:04:01.765888
+AAPL

# 2021-12-30 21:04:04.859463
+MSFT

# 2021-12-30 21:14:51.724512
+AAPL

# 2021-12-30 21:14:54.346682
+MSFT

# 2021-12-30 21:15:14.627662
+AAPL

# 2021-12-30 21:15:19.111552
+MSFT

# 2021-12-30 21:16:05.412891
+AAPL

# 2021-12-30 21:16:08.409010
+MSFT

# 2021-12-30 21:17:37.020824
+AAPL

# 2021-12-30 22:24:06.850011
+MSFT

# 2021-12-30 22:29:19.598194
+AAPL

# 2021-12-30 22:29:22.630208
+MSFT

# 2021-12-30 22:30:39.347292
+AAPL

# 2021-12-30 22:30:42.699520
+MSFT

# 2021-12-30 22:32:19.546876
+AAPL

# 2021-12-30 22:32:22.579448
+MSFT

# 2021-12-30 22:33:43.117112
+AAPL

# 2021-12-30 22:33:45.854237
+MSFT

# 2021-12-30 22:34:08.065328
+AAPL

# 2021-12-30 22:34:10.887615
+MSFT

# 2021-12-30 22:35:18.651804
+TSLA

# 2021-12-30 22:36:47.010635
+MSFT

# 2021-12-30 22:36:49.780876
+AAPL

# 2021-12-30 22:37:10.697957
+MSFT

# 2021-12-30 22:39:25.955060
+AAPL

# 2021-12-30 22:39:29.073080
+MSFT

# 2021-12-30 22:40:06.413000
+AAPL

# 2021-12-30 22:40:09.061982
+MSFT

# 2021-12-30 22:46:02.858782
+AAPL

# 2021-12-30 22:46:05.733293
+MSFT

# 2021-12-30 22:46:31.331292
+AAPL

# 2021-12-30 22:46:33.959151
+MSFT

# 2021-12-30 22:54:08.138091
+AAPL

# 2021-12-30 22:54:43.344129
+MSFT

# 2021-12-30 22:55:23.571840
+SPY

# 2021-12-30 22:55:27.775582
+TSLA

# 2021-12-30 22:55:51.401846
+AAPL

# 2021-12-30 22:55:54.956697
+MSFT

# 2021-12-30 22:58:20.238527
+SPY

# 2021-12-30 22:58:23.199626
+TSLA

# 2021-12-30 22:59:14.014846
+SPY

# 2021-12-30 22:59:18.039817
+TSLA

# 2021-12-30 22:59:54.646214
+RBLX

# 2021-12-30 23:00:03.656830
+MU

# 2021-12-30 23:02:29.247960
+AAPL

# 2021-12-30 23:02:31.982964
+MSFT

# 2021-12-30 23:03:33.561368
+AAPL

# 2021-12-30 23:03:36.307936
+MSFT

# 2021-12-30 23:03:52.938877
+AAPL

# 2021-12-30 23:03:58.154700
+MSFT

# 2021-12-30 23:14:15.925454
+AAPL

# 2021-12-30 23:14:18.972256
+MSFT


IBAPI.py
from ibapi.client import EClient
from ibapi.wrapper import EWrapper 
from ContractSamples import *
from ibapi.common import * # @UnusedWildImport
from ibapi.utils import * # @UnusedWildImport
from ibapi.contract import (Contract, ContractDetails, DeltaNeutralContract)
from ibapi.order import Order
from ibapi.order_state import OrderState
from ibapi.execution import Execution
from ibapi.ticktype import * # @UnusedWildImport
from ibapi.commission_report import CommissionReport
from CsvUtility import CsvUtility

import datetime
import time
import threading
import os

class IBAPI(EWrapper, EClient):
    def __init__(self):
        EClient.__init__(self, self)
        headers = ['conId', 'symbol', 'secType', 'primExchange', 'currency', 'derivativeSecTypes']
        self.csvClient = CsvUtility(headers)

    def historicalData(self, reqId:int, bar: BarData):
        # print("HistoricalData. ReqId:", reqId, "BarData.", bar)
        #rowData = str(bar.date) + "," + str(bar.open) + "," +  str(bar.high) + "," +  str(bar.low) + "," +  str(bar.close) + "," +  str(bar.volume) + "," +  str(bar.average) + "," +  str(bar.barCount)
        barData = [bar.date, bar.open, bar.high, bar.low, bar.close, bar.volume, bar.average, bar.barCount]
        self.csvClient.insertRow(barData)

    def historicalDataEnd(self, reqId: int, start: str, end: str):
        super().historicalDataEnd(reqId, start, end)
        print("HistoricalDataEnd. ReqId:", reqId, "from", start, "to", end)

    def historicalDataUpdate(self, reqId: int, bar: BarData):
        print("HistoricalDataUpdate. ReqId:", reqId, "BarData.", bar)

    def symbolSamples(self, reqId: int, contractDescriptions: ListOfContractDescription):
        super().symbolSamples(reqId, contractDescriptions)
        print("Symbol Samples. Request Id: ", reqId)
        
        headers = ['conId', 'symbol', 'secType', 'primExchange', 'currency', 'derivativeSecTypes']
        csvUtility = CsvUtility(path=os.getcwd()+"\\choices.csv", headers=headers)

        for contractDescription in contractDescriptions:
            derivSecTypes = ""
            for derivSecType in contractDescription.derivativeSecTypes:
                derivSecTypes += derivSecType
            derivSecTypes += " "
            currentRow = [
                contractDescription.contract.conId,
                contractDescription.contract.symbol,
                contractDescription.contract.secType,
                contractDescription.contract.primaryExchange,
                contractDescription.contract.currency, derivSecTypes
            ]
            csvUtility.insertRow(currentRow)
            # print("Contract: conId:%s, symbol:%s, secType:%s primExchange:%s, "
            #         "currency:%s, derivativeSecTypes:%s" % (
            #     contractDescription.contract.conId,
            #     contractDescription.contract.symbol,
            #     contractDescription.contract.secType,
            #     contractDescription.contract.primaryExchange,
            #     contractDescription.contract.currency, derivSecTypes))

class Operations():
    def __init__(self, host, port, clientId):
        self.client = IBAPI()
        self.client.connect(host, port, clientId)
        self.api_thread = threading.Thread(target = self.runLoop, daemon=True)
        self.api_thread.start()
        time.sleep(1)

    def runLoop(self):
        self.client.run()
    
    def getHistoricalMarketData(self, reqId, contract, queryTime, durationString, barSize, historicalDataType):
        self.client.reqHistoricalData(reqId, contract, queryTime, durationString, barSize, historicalDataType, 1, 1, False, [])

    def stockSearch(self, symbol):
        self.client.reqMatchingSymbols(1, symbol)
        
# x.getHistoricalMarketData("symbol")

# time.sleep(100) #Sleep interval to allow time for incoming price data

appletest.py
from ibapi.client import EClient
from ibapi.wrapper import EWrapper
from ibapi.contract import Contract

import threading
import time


class IBapi(EWrapper, EClient):
	def __init__(self):
		EClient.__init__(self, self)
	def tickPrice(self, reqId, tickType, price, attrib):
		if tickType == 2 and reqId == 1:
			print('The current ask price is: ', price)

def run_loop():
	app.run()

app = IBapi()
app.connect('127.0.0.1', 4001, 123)

#Start the socket in a thread
api_thread = threading.Thread(target=run_loop, daemon=True)
api_thread.start()

time.sleep(1) #Sleep interval to allow time for connection to server

#Create contract object
apple_contract = Contract()
apple_contract.symbol = 'AAPL'
apple_contract.secType = 'STK'
apple_contract.exchange = 'SMART'
apple_contract.currency = 'USD'

#Request Market Data
app.reqMktData(1, apple_contract, '', False, False, [])

time.sleep(100) #Sleep interval to allow time for incoming price data
app.disconnect()

CsvUtility.py
from HistoricalDataEnums import *

import csv
import time

class CsvUtility:
    def __init__(self, headers, path=None):
        self.path = path
        self.headers = headers

        if path is not None:
            with open(path, 'w', encoding='UTF8', newline='') as file:
                csv.writer(file).writerow(self.headers)

    def insertRow(self, row):
        with open(self.path, 'a', encoding='UTF8', newline='') as file:
            csv.writer(file).writerow(row)
    
    def setPath(self, headers, path):
        with open(path, 'w', encoding='UTF8', newline='') as file:
                csv.writer(file).writerow(headers)
        self.path = path

    def readCsvToList(self, path, type):
        with open(path, 'r', encoding='UTF8') as file:
            reader = csv.reader(file)
            headers = next(reader)

            resultsDictionary = {}
            
            if type == HistoricalDataEnums.SYMBOL_LOOKUP_DATA.value:
                for num, line in enumerate(reader):
                    if line[3] == "NASDAQ.NMS":
                        line[3] = "SMART"
                    resultsDictionary[num] = line
            elif type == HistoricalDataEnums.RAW_EQUITY_DATA.value:
                for num, line in enumerate(reader):
                    resultsDictionary[num] = line

            return [headers, resultsDictionary]
    
    def validateTwoDatasetDatesAndReturnCloseValues(self, dataSetOne, dataSetTwo, closeIdx):
        dataSetOneCloses = []
        dataSetTwoCloses = []

        lastDataOneValue = None
        lastDataTwoValue = None

        if len(dataSetOne) != len(dataSetTwo):
            return [None, None]
        for key in dataSetOne:
            if dataSetOne[key][0] != dataSetTwo[key][0]:
                return [None, None]
            if lastDataOneValue is None and lastDataTwoValue is None:
                lastDataOneValue = float(dataSetOne[key][closeIdx])
                lastDataTwoValue = float(dataSetTwo[key][closeIdx])
            else:
                currentValue = float(dataSetOne[key][closeIdx])
                percentChange = self.percentChangeTwoValues(lastDataOneValue, currentValue)
                dataSetOneCloses.append(percentChange)
                lastDataOneValue = currentValue

                currentValue = float(dataSetTwo[key][closeIdx])
                percentChange = self.percentChangeTwoValues(lastDataTwoValue, currentValue)
                dataSetTwoCloses.append(percentChange)
                lastDataTwoValue = currentValue

        return [dataSetOneCloses, dataSetTwoCloses]
    
    def percentChangeTwoValues(self, valueOne, valueTwo):
        return (valueTwo-valueOne)/valueOne

HistoricalDataEnums.py
from enum import Enum
import datetime

class HistoricalDataEnums(Enum):
    DURATION_ONESECOND = "1 S"
    DURATION_ONEDAY   = "1 D"
    DURATION_ONEWEEK   = "1 W"
    DURATION_ONEMONTH  = "1 M"
    DURATION_ONEYEAR   = "1 Y"
    DURATION_TWOYEARS  = "2 Y"
    DURATION_FIVEYEARS = "5 Y"

    BAR_SIZE_ONESECOND = "1 secs"
    BAR_SIZE_FIVESECONDS = "5 secs"	
    BAR_SIZE_TENSECONDS = "10 secs"	
    BAR_SIZE_FIFTEENSECONDS = "15 secs"	
    BAR_SIZE_THIRTYSECONDS = "30 secs"
    BAR_SIZE_ONEMINUTE = "1 min"	
    BAR_SIZE_TWOMINUTES = "2 mins"	
    BAR_SIZE_THREEMINUTES = "3 mins"
    BAR_SIZE_FIVEMINUTES = "5 mins"	
    BAR_SIZE_TENMINUTES = "10 mins"	
    BAR_SIZE_FIFTEENMINUTES = "15 mins"	
    BAR_SIZE_TWENTYMINUTES = "20 mins"	
    BAR_SIZE_THIRTYMINUTES = "30 mins"
    BAR_SIZE_ONEHOUR = "1 hour"	
    BAR_SIZE_TWOHOURS = "2 hours"	
    BAR_SIZE_THREEHOURS = "3 hours"	
    BAR_SIZE_FOURHOURS = "4 hours"	
    BAR_SIZE_EIGHTHOURS = "8 hours"
    BAR_SIZE_ONEDAY = "1 day"
    BAR_SIZE_ONEWEEK = "1 week"
    BAR_SIZE_ONEMONTH = "1 month"

    HISTORICAL_DATA_TYPE_TRADES = "TRADES"
    HISTORICAL_DATA_TYPE_MIDPOINT = "MIDPOINT"
    HISTORICAL_DATA_TYPE_BID = "BID"
    HISTORICAL_DATA_TYPE_ASK = "ASK"
    HISTORICAL_DATA_TYPE_BIDASK = "BID_ASK"
    HISTORICAL_DATA_TYPE_ADJUSTEDLAST = "ADJUSTED_LAST"
    HISTORICAL_DATA_TYPE_HISTORICALVOLATILITY = "HISTORICAL_VOLATILITY"
    HISTORICAL_DATA_TYPE_OPTIONIMPLIEDVOLATILITY = "OPTION_IMPLIED_VOLATILITY"
    HISTORICAL_DATA_TYPE_REBATERATE = "REBATE_RATE"
    HISTORICAL_DATA_TYPE_FEERATE = "FEE_RATE"
    HISTORICAL_DATA_TYPE_YIELDBID = "YIELD_BID"
    HISTORICAL_DATA_TYPE_YIELDASK = "YIELD_ASK"
    HISTORICAL_DATA_TYPE_YIELDBIDASK = "YIELD_BID_ASK"
    HISTORICAL_DATA_TYPE_YIELDLAST = "YIELD_LAST"

    SYMBOL_LOOKUP_DATA = "SYMBOL_LOOKUP_DATA"
    RAW_EQUITY_DATA = "RAW_EQUITY_DATA"

    QUERY_TIME = (datetime.datetime.today() - datetime.timedelta(days=1)).strftime("%Y%m%d %H:%M:%S")
    

ContractSamples.py
"""
Copyright (C) 2019 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable.
"""

from ibapi.contract import * # @UnusedWildImport


class ContractSamples:

    """ Usually, the easiest way to define a Stock/CASH contract is through 
    these four attributes.  """

    @staticmethod
    def EurGbpFx():
        #! [cashcontract]
        contract = Contract()
        contract.symbol = "EUR"
        contract.secType = "CASH"
        contract.currency = "GBP"
        contract.exchange = "IDEALPRO"
        #! [cashcontract]
        return contract


    @staticmethod
    def Index():
        #! [indcontract]
        contract = Contract()
        contract.symbol = "DAX"
        contract.secType = "IND"
        contract.currency = "EUR"
        contract.exchange = "DTB"
        #! [indcontract]
        return contract


    @staticmethod
    def CFD():
        #! [cfdcontract]
        contract = Contract()
        contract.symbol = "IBDE30"
        contract.secType = "CFD"
        contract.currency = "EUR"
        contract.exchange = "SMART"
        #! [cfdcontract]
        return contract


    @staticmethod
    def EuropeanStock():
        contract = Contract()
        contract.symbol = "BMW"
        contract.secType = "STK"
        contract.currency = "EUR"
        contract.exchange = "SMART"
        contract.primaryExchange = "IBIS"
        return contract

    @staticmethod
    def EuropeanStock2():
        contract = Contract()
        contract.symbol = "NOKIA"
        contract.secType = "STK"
        contract.currency = "EUR"
        contract.exchange = "SMART"
        contract.primaryExchange = "HEX"
        return contract

    @staticmethod
    def OptionAtIse():
        contract = Contract()
        contract.symbol = "COF"
        contract.secType = "OPT"
        contract.currency = "USD"
        contract.exchange = "ISE"
        contract.lastTradeDateOrContractMonth = "20190315"
        contract.right = "P"
        contract.strike = 105
        contract.multiplier = "100"
        return contract


    @staticmethod
    def BondWithCusip():
            #! [bondwithcusip]
            contract = Contract()
            # enter CUSIP as symbol
            contract.symbol= "912828C57"
            contract.secType = "BOND"
            contract.exchange = "SMART"
            contract.currency = "USD"
            #! [bondwithcusip]
            return contract


    @staticmethod
    def Bond():
            #! [bond]
            contract = Contract()
            contract.conId = 15960357
            contract.exchange = "SMART"
            #! [bond]
            return contract


    @staticmethod
    def MutualFund():
            #! [fundcontract]
            contract = Contract()
            contract.symbol = "VINIX"
            contract.secType = "FUND"
            contract.exchange = "FUNDSERV"
            contract.currency = "USD"
            #! [fundcontract]
            return contract


    @staticmethod
    def Commodity():
            #! [commoditycontract]
            contract = Contract()
            contract.symbol = "XAUUSD"
            contract.secType = "CMDTY"
            contract.exchange = "SMART"
            contract.currency = "USD"
            #! [commoditycontract]
            return contract
    

    @staticmethod
    def USStock():
        #! [stkcontract]
        contract = Contract()
        contract.symbol = "IBKR"
        contract.secType = "STK"
        contract.currency = "USD"
        #In the API side, NASDAQ is always defined as ISLAND in the exchange field
        contract.exchange = "ISLAND"
        #! [stkcontract]
        return contract


    @staticmethod
    def USStockWithPrimaryExch():
        #! [stkcontractwithprimary]
        contract = Contract()
        contract.symbol = "MSFT"
        contract.secType = "STK"
        contract.currency = "USD"
        contract.exchange = "SMART"
        #Specify the Primary Exchange attribute to avoid contract ambiguity 
        #(there is an ambiguity because there is also a MSFT contract with primary exchange = "AEB")
        contract.primaryExchange = "ISLAND"
        #! [stkcontractwithprimary]
        return contract

            
    @staticmethod
    def USStockAtSmart():
        contract = Contract()
        contract.symbol = "IBM"
        contract.secType = "STK"
        contract.currency = "USD"
        contract.exchange = "SMART"
        return contract


    @staticmethod
    def USOptionContract():
        #! [optcontract_us]
        contract = Contract()
        contract.symbol = "GOOG"
        contract.secType = "OPT"
        contract.exchange = "SMART"
        contract.currency = "USD"
        contract.lastTradeDateOrContractMonth = "20190315"
        contract.strike = 1180
        contract.right = "C"
        contract.multiplier = "100"
        #! [optcontract_us]
        return contract


    @staticmethod
    def OptionAtBOX():
        #! [optcontract]
        contract = Contract()
        contract.symbol = "GOOG"
        contract.secType = "OPT"
        contract.exchange = "BOX"
        contract.currency = "USD"
        contract.lastTradeDateOrContractMonth = "20190315"
        contract.strike = 1180
        contract.right = "C"
        contract.multiplier = "100"
        #! [optcontract]
        return contract


    """ Option contracts require far more information since there are many 
    contracts having the exact same attributes such as symbol, currency, 
    strike, etc. This can be overcome by adding more details such as the 
    trading class"""

    @staticmethod
    def OptionWithTradingClass():
        #! [optcontract_tradingclass]
        contract = Contract()
        contract.symbol = "SANT"
        contract.secType = "OPT"
        contract.exchange = "MEFFRV"
        contract.currency = "EUR"
        contract.lastTradeDateOrContractMonth = "20190621"
        contract.strike = 7.5
        contract.right = "C"
        contract.multiplier = "100"
        contract.tradingClass = "SANEU"
        #! [optcontract_tradingclass]
        return contract


    """ Using the contract's own symbol (localSymbol) can greatly simplify a
    contract description """

    @staticmethod
    def OptionWithLocalSymbol():
        #! [optcontract_localsymbol]
        contract = Contract()
        #Watch out for the spaces within the local symbol!
        contract.localSymbol = "C DBK  DEC 20  1600"
        contract.secType = "OPT"
        contract.exchange = "DTB"
        contract.currency = "EUR"
        #! [optcontract_localsymbol]
        return contract

    """ Dutch Warrants (IOPTs) can be defined using the local symbol or conid 
    """

    @staticmethod
    def DutchWarrant():
        #! [ioptcontract]
        contract = Contract()
        contract.localSymbol = "B881G"
        contract.secType = "IOPT"
        contract.exchange = "SBF"
        contract.currency = "EUR"
        #! [ioptcontract]
        return contract

    """ Future contracts also require an expiration date but are less
    complicated than options."""

    @staticmethod
    def SimpleFuture():
        #! [futcontract]
        contract = Contract()
        contract.symbol = "ES"
        contract.secType = "FUT"
        contract.exchange = "GLOBEX"
        contract.currency = "USD"
        contract.lastTradeDateOrContractMonth = "201903"
        #! [futcontract]
        return contract


    """Rather than giving expiration dates we can also provide the local symbol
    attributes such as symbol, currency, strike, etc. """

    @staticmethod
    def FutureWithLocalSymbol():
        #! [futcontract_local_symbol]
        contract = Contract()
        contract.secType = "FUT"
        contract.exchange = "GLOBEX"
        contract.currency = "USD"
        contract.localSymbol = "ESU6"
        #! [futcontract_local_symbol]
        return contract


    @staticmethod
    def FutureWithMultiplier():
        #! [futcontract_multiplier]
        contract = Contract()
        contract.symbol = "DAX"
        contract.secType = "FUT"
        contract.exchange = "DTB"
        contract.currency = "EUR"
        contract.lastTradeDateOrContractMonth = "201903"
        contract.multiplier = "5"
        #! [futcontract_multiplier]
        return contract


    """ Note the space in the symbol! """

    @staticmethod
    def WrongContract():
        contract = Contract()
        contract.symbol = " IJR "
        contract.conId = 9579976
        contract.secType = "STK"
        contract.exchange = "SMART"
        contract.currency = "USD"
        return contract

    @staticmethod
    def FuturesOnOptions():
        #! [fopcontract]
        contract = Contract()
        contract.symbol = "ES"
        contract.secType = "FOP"
        contract.exchange = "GLOBEX"
        contract.currency = "USD"
        contract.lastTradeDateOrContractMonth = "20190315"
        contract.strike = 2900
        contract.right = "C"
        contract.multiplier = "50"
        #! [fopcontract]
        return contract


    """ It is also possible to define contracts based on their ISIN (IBKR STK
    sample). """

    @staticmethod
    def ByISIN():
        contract = Contract()
        contract.secIdType = "ISIN"
        contract.secId = "US45841N1072"
        contract.exchange = "SMART"
        contract.currency = "USD"
        contract.secType = "STK"
        return contract


    """ Or their conId (EUR.uSD sample).
    Note: passing a contract containing the conId can cause problems if one of 
    the other provided attributes does not match 100% with what is in IB's 
    database. This is particularly important for contracts such as Bonds which 
    may change their description from one day to another.
    If the conId is provided, it is best not to give too much information as
    in the example below. """

    @staticmethod
    def ByConId():
        contract = Contract()
        contract.secType = "CASH"
        contract.conId = 12087792
        contract.exchange = "IDEALPRO"
        return contract


    """ Ambiguous contracts are great to use with reqContractDetails. This way
    you can query the whole option chain for an underlying. Bear in mind that
    there are pacing mechanisms in place which will delay any further responses
    from the TWS to prevent abuse. """

    @staticmethod
    def OptionForQuery():
        #! [optionforquery]
        contract = Contract()
        contract.symbol = "FISV"
        contract.secType = "OPT"
        contract.exchange = "SMART"
        contract.currency = "USD"
        #! [optionforquery]
        return contract


    @staticmethod
    def OptionComboContract():
        #! [bagoptcontract]
        contract = Contract()
        contract.symbol = "DBK"
        contract.secType = "BAG"
        contract.currency = "EUR"
        contract.exchange = "DTB"

        leg1 = ComboLeg()
        leg1.conId = 317960956 #DBK JUN 21 2019 C
        leg1.ratio = 1
        leg1.action = "BUY"
        leg1.exchange = "DTB"

        leg2 = ComboLeg()
        leg2.conId = 334216780  #DBK MAR 15 2019 C
        leg2.ratio = 1
        leg2.action = "SELL"
        leg2.exchange = "DTB"

        contract.comboLegs = []
        contract.comboLegs.append(leg1)
        contract.comboLegs.append(leg2)
        #! [bagoptcontract]
        return contract


    """ STK Combo contract
    Leg 1: 43645865 - IBKR's STK
    Leg 2: 9408 - McDonald's STK """

    @staticmethod
    def StockComboContract():
        #! [bagstkcontract]
        contract = Contract()
        contract.symbol = "IBKR,MCD"
        contract.secType = "BAG"
        contract.currency = "USD"
        contract.exchange = "SMART"

        leg1 = ComboLeg()
        leg1.conId = 43645865#IBKR STK
        leg1.ratio = 1
        leg1.action = "BUY"
        leg1.exchange = "SMART"

        leg2 = ComboLeg()
        leg2.conId = 9408#MCD STK
        leg2.ratio = 1
        leg2.action = "SELL"
        leg2.exchange = "SMART"

        contract.comboLegs = []
        contract.comboLegs.append(leg1)
        contract.comboLegs.append(leg2)
        #! [bagstkcontract]
        return contract


    """ CBOE Volatility Index Future combo contract """

    @staticmethod
    def FutureComboContract():
        #! [bagfutcontract]
        contract = Contract()
        contract.symbol = "VIX"
        contract.secType = "BAG"
        contract.currency = "USD"
        contract.exchange = "CFE"

        leg1 = ComboLeg()
        leg1.conId = 326501438 # VIX FUT 201903
        leg1.ratio = 1
        leg1.action = "BUY"
        leg1.exchange = "CFE"

        leg2 = ComboLeg()
        leg2.conId = 323072528 # VIX FUT 2019049
        leg2.ratio = 1
        leg2.action = "SELL"
        leg2.exchange = "CFE"

        contract.comboLegs = []
        contract.comboLegs.append(leg1)
        contract.comboLegs.append(leg2)
        #! [bagfutcontract]
        return contract

    @staticmethod
    def SmartFutureComboContract():
        #! [smartfuturespread]
        contract = Contract()
        contract.symbol = "WTI" # WTI,COIL spread. Symbol can be defined as first leg symbol ("WTI") or currency ("USD")
        contract.secType = "BAG"
        contract.currency = "USD"
        contract.exchange = "SMART"

        leg1 = ComboLeg()
        leg1.conId = 55928698 # WTI future June 2017
        leg1.ratio = 1
        leg1.action = "BUY"
        leg1.exchange = "IPE"

        leg2 = ComboLeg()
        leg2.conId = 55850663 # COIL future June 2017
        leg2.ratio = 1
        leg2.action = "SELL"
        leg2.exchange = "IPE"

        contract.comboLegs = []
        contract.comboLegs.append(leg1)
        contract.comboLegs.append(leg2)
        #! [smartfuturespread]
        return contract

    @staticmethod
    def InterCmdtyFuturesContract():
        #! [intcmdfutcontract]
        contract = Contract()
        contract.symbol = "CL.BZ" #symbol is 'local symbol' of intercommodity spread. 
        contract.secType = "BAG"
        contract.currency = "USD"
        contract.exchange = "NYMEX"

        leg1 = ComboLeg()
        leg1.conId = 47207310 #CL Dec'16 @NYMEX
        leg1.ratio = 1
        leg1.action = "BUY"
        leg1.exchange = "NYMEX"

        leg2 = ComboLeg()
        leg2.conId = 47195961 #BZ Dec'16 @NYMEX
        leg2.ratio = 1
        leg2.action = "SELL"
        leg2.exchange = "NYMEX"

        contract.comboLegs = []
        contract.comboLegs.append(leg1)
        contract.comboLegs.append(leg2)
        #! [intcmdfutcontract]
        return contract


    @staticmethod
    def NewsFeedForQuery():
        #! [newsfeedforquery]
        contract = Contract()
        contract.secType = "NEWS"
        contract.exchange = "BRFG" #Briefing Trader
        #! [newsfeedforquery]
        return contract


    @staticmethod
    def BRFGbroadtapeNewsFeed():
        #! [newscontractbt]
        contract = Contract()
        contract.symbol  = "BRFG:BRFG_ALL"
        contract.secType = "NEWS"
        contract.exchange = "BRFG"
        #! [newscontractbt]
        return contract


    @staticmethod
    def DJNLbroadtapeNewsFeed():
        #! [newscontractbz]
        contract = Contract()
        contract.symbol = "DJNL:DJNL_ALL"
        contract.secType = "NEWS"
        contract.exchange = "DJNL"
        #! [newscontractbz]
        return contract


    @staticmethod
    def DJTOPbroadtapeNewsFeed():
        #! [newscontractfly]
        contract = Contract()
        contract.symbol  = "DJTOP:ASIAPAC"
        contract.secType = "NEWS"
        contract.exchange = "DJTOP"
        #! [newscontractfly]
        return contract


    @staticmethod
    def BRFUPDNbroadtapeNewsFeed():
        #! [newscontractmt]
        contract = Contract()
        contract.symbol = "BRFUPDN:BRF_ALL"
        contract.secType = "NEWS"
        contract.exchange = "BRFUPDN"
        #! [newscontractmt]
        return contract

    @staticmethod
    def ContFut():
        #! [continuousfuturescontract]
        contract = Contract()
        contract.symbol = "ES"
        contract.secType = "CONTFUT"
        contract.exchange = "GLOBEX"
        #! [continuousfuturescontract]
        return contract

    @staticmethod
    def ContAndExpiringFut():
        #! [contandexpiringfut]
        contract = Contract()
        contract.symbol = "ES"
        contract.secType = "FUT+CONTFUT"
        contract.exchange = "GLOBEX"
        #! [contandexpiringfut]
        return contract

    @staticmethod
    def JefferiesContract():
        #! [jefferies_contract]
        contract = Contract()
        contract.symbol = "AAPL"
        contract.secType = "STK"
        contract.exchange = "JEFFALGO"
        contract.currency = "USD"
        #! [jefferies_contract]
        return contract

    @staticmethod
    def CSFBContract():
        #! [csfb_contract]
        contract = Contract()
        contract.symbol = "IBKR"
        contract.secType = "STK"
        contract.exchange = "CSFBALGO"
        contract.currency = "USD"
        #! [csfb_contract]
        return contract

    @staticmethod
    def USStockCFD():
        # ! [usstockcfd_conract]
        contract = Contract();
        contract.symbol = "IBM";
        contract.secType = "CFD";
        contract.currency = "USD";
        contract.exchange = "SMART";
        # ! [usstockcfd_conract]
        return contract;

    @staticmethod
    def EuropeanStockCFD():
        # ! [europeanstockcfd_contract]
        contract = Contract();
        contract.symbol = "BMW";
        contract.secType = "CFD";
        contract.currency = "EUR";
        contract.exchange = "SMART";
        # ! [europeanstockcfd_contract]
        return contract;

    @staticmethod
    def CashCFD():
        # ! [cashcfd_contract]
        contract = Contract();
        contract.symbol = "EUR";
        contract.secType = "CFD";
        contract.currency = "USD";
        contract.exchange = "SMART";
        # ! [cashcfd_contract]
        return contract;

    @staticmethod
    def QBAlgoContract():
        # ! [qbalgo_contract]
        contract = Contract()
        contract.symbol = "ES";
        contract.secType = "FUT";
        contract.exchange = "QBALGO";
        contract.currency = "USD";
        contract.lastTradeDateOrContractMonth = "202003";
        # ! [qbalgo_contract]
        return contract;  

def Test():
    from ibapi.utils import ExerciseStaticMethods
    ExerciseStaticMethods(ContractSamples)


if "__main__" == __name__:
    Test()

